[wip] Bootloader drivers

algernon opened the pull request:
The various keyboard plugins implemented a way to jump to the bootloader on their own. However, this is a bootloader-specific functionality, and as such, it makes sense to have a common implementation.

This patch does just that.

algernon commented:
I'm a bit torn about the API here. Another idea would be to have one `kaleidoscope::driver::Bootloader` class, with a templated `resetDevice` method. Thus, hardware plugins would subclass the bootloader driver, and call whichever `resetDevice` is appropriate. We can also template the whole class, because at run-time, we only support one bootloader.

This is not tied to the MCU, and it needs to be changeable within a hardware plugin, because we support boards that can have different bootloaders (Atreus, ErgoDox) depending on build.

Feedback on the API, or on the idea itself would be most appreciated.

noseglasses commented:
This could probably be solved more versatile by using composition. Provide a library of bootloader classes and let the hardware classes choose the appropriate one to become a member. 

What would be the advantage of using templates here? One that I could imagine is that templates enable conditional compilation. But I am not sure if that is required here.

algernon commented:
> Provide a library of bootloader classes and let the hardware classes choose the appropriate one to become a member.

That's very similar to what is done in this PR: there are bootloader classes with static `resetDevice` members. Hardware plugins have their `resetDevice()` method call the appropriate bootloader's `resetDevice`.

Like in the MCU case, I want to avoid having a `getBootloader()` method or similar.

> What would be the advantage of using templates here? One that I could imagine is that templates enable conditional compilation. But I am not sure if that is required here.

Templates would make `resetDevice` invocation a little shorter:

```c++
void resetDevice() {
  kaleidoscope::driver::Bootloader::resetDevice<Caterina>();
}
```

or:

```c++
void resetDevice() {
  kaleidoscope::driver::Bootloader<Caterina>::resetDevice();
}
```

or, if using inheritance:

```c++
class SomeHardware : public kaleidoscope::Hardware, protected kaleidoscope::driver::Bootloader<Caterina> {
  // ...
};
```

The goal is to make it easy for hardware plugins to pull in the right thing. Neither these, nor the MCU helpers will be exposed to other plugins, let alone end-users (only via wrappers in the hardware class at best).

algernon commented:
Following the MCU example:

```c++
class SomeHardware: public kaleidoscope::Hardware {
 public:
  void resetDevice {
    bootloader_.resetDevice();
  }
 protected:
  kaleidoscope::driver::bootloader::Caterina bootloader_;
};
```

This looks fine.There's a bit of boilerplate required to wrap the bootloader's `resetDevice()`, but we have similar wrappers all around the place anyway.

algernon commented:
Updated.

There are some size increases I want to investigate though.

gedankenexperimenter commented on src/kaleidoscope/driver/Bootloader.h:
```diff
@@ -0,0 +1,124 @@
+    // Set a watchdog timer
+    wdt_enable(WDTO_120MS);
+
+    while (1) {} // This infinite loop ensures nothing else
```
One style of infinite loop here (personally, I prefer `while (true)`, but whatever…).

noseglasses commented:
Nice.

Allow me one general remark. There's a reason why I suggest having instances of classes MCU or Bootloaders and let the hardware class call their non-static members. The reason is extensibility.

```cpp
struct A {
   static void doSomething() { ... }
};
...
// Which approach is more extensible?
A a; // lives beyond the calling scope of doSomething()
//...
a.doSomething();

// or
A::doSomething();
```

Even though this approach requires an instance of `class A` which costs at minimum one byte, using a member instance and call its member functions will enable adding state to a formerly stateless classes later on. As the caller side does not need to know whether a method is static or not, IMHO it is always recommended to prefer the `a.doSomething()` approach. It comes with zero additional overhead in its initial configuration. Later on, the static method can be made non-static. This will then add additional overhead for passing the `this` pointer, but for good reasons.

In the past I have often experienced that a design was intended to be stateless in the first place and then turned out to require some state variables later on. When writing `A::doSomething()` instead of `a.doSomething()`, the caller side must be modified when the implementation changes.

This is just a general remark, I am not sure if this ever applies to MCU or Bootloaders. It just proved to be useful to restrict the use of global and static member functions to a minimum.

The same can be generalized to Kaleidoscope as a whole. There are quite a few classes that mostly carry static member functions which are called as `a_namespace::a_class::a_static_method(...)'. In the absence of template meta-programming (where juggling with types instead of objects is common) this approach is non-optimal for the reasons stated above.

gedankenexperimenter commented on src/kaleidoscope/driver/Bootloader.h:
```diff
@@ -0,0 +1,124 @@
+  static inline void resetDevice() {
+    reset_key = BOOTLOADER_RESET_KEY;
+    wdt_enable(WDTO_250MS);
+    for (;;);
```
Another style of infinite loop here, in the same file.

algernon replied:
Good catch. These were all copied from different - usually C - sources, hence the different styles. I'll update the PR to use `while (true)` instead. Thanks!

noseglasses commented:
Just wanted to add that the additional byte for a class instance only applies for global or local variables, not for class members.

algernon commented:
> `a::something()` vs `a.something()`

Current code uses `a.something()`, where a is an instance of the bootloader class.

Regarding size increase, I'm seeing an increase of  ~100 bytes in some cases, not just a few which I would have expected.

algernon commented:
Size differences:

| Board | driver/mcu | driver/bootloader |
|---------|--|--|
| Splitography | 11016 / 648 | 11448 / 681 |

algernon commented:
Aha, ok, I see what's up, at least in the Splitography case. Need to move some of the stuff to `.cpp`.

algernon commented:
Ok, things look better now:

```diff
--- SIZES.mcu   2019-03-16 19:25:03.361885795 +0100
+++ SIZES.bootloader    2019-03-16 19:23:52.666347261 +0100
@@ -1,7 +1,7 @@
 Building examples/Devices/Atreus/Atreus 0.0.0 into /tmp/kaleidoscope-algernon/sketch/3541210-Atreus.ino/output...                                
 - Size: firmware/Devices/Atreus/Atreus-0.0.0.elf
-  - Program:   11742 bytes (41.0% Full)
-  - Data:        667 bytes (26.1% Full)
+  - Program:   11732 bytes (40.9% Full)
+  - Data:        666 bytes (26.0% Full)

 Building examples/Devices/ErgoDox/ErgoDox 0.0.0 into /tmp/kaleidoscope-algernon/sketch/3539384-ErgoDox.ino/output...                             
 - Size: firmware/Devices/ErgoDox/ErgoDox-0.0.0.elf
@@ -10,13 +10,13 @@

 Building examples/Devices/KBD4x/KBD4x 0.0.0 into /tmp/kaleidoscope-algernon/sketch/3540870-KBD4x.ino/output...                                   
 - Size: firmware/Devices/KBD4x/KBD4x-0.0.0.elf
-  - Program:   11708 bytes (40.8% Full)
-  - Data:        689 bytes (26.9% Full)
+  - Program:   11698 bytes (40.8% Full)
+  - Data:        688 bytes (26.9% Full)

 Building examples/Devices/Keyboardio/Kaleidoscope/Kaleidoscope 0.0.0 into /tmp/kaleidoscope-algernon/sketch/3539388-Kaleidoscope.ino/output...   
 - Size: firmware/Devices/Keyboardio/Kaleidoscope/Kaleidoscope-0.0.0.elf
   - Program:   18204 bytes (63.5% Full)
-  - Data:       1215 bytes (47.5% Full)
+  - Data:       1216 bytes (47.5% Full)

 Building examples/Devices/Planck/Planck 0.0.0 into /tmp/kaleidoscope-algernon/sketch/3539406-Planck.ino/output...                                
 - Size: firmware/Devices/Planck/Planck-0.0.0.elf
@@ -26,5 +26,5 @@
 Building examples/Devices/Splitography/Splitography 0.0.0 into /tmp/kaleidoscope-algernon/sketch/3539409-Splitography.ino/output...              
 - Size: firmware/Devices/Splitography/Splitography-0.0.0.elf
   - Program:   11016 bytes (38.4% Full)
-  - Data:        648 bytes (25.3% Full)
+  - Data:        649 bytes (25.4% Full)
```

In the Atreus (HalfKay), and KBD4x (DFU) cases, we're 10 bytes PROGMEM  and one byte RAM less. Model01 (Caterina) and Splitography (DFU) saw one byte RAM more.

This doesn't make much sense to me, will have to compare the assembly to see what's different. Mind you, infinite loops were all changed to `while (true)` from `while(1)` and `for(;;)`, that might be a contributing factor.

algernon commented:
The loops don't matter, and looks like things are properly compiled. Perhaps its DFU not being inlined? Not sure about the Atreus case though.

Will look at assembly later tonight. The size differences are ok now, I just want to make sure everything still works.

gedankenexperimenter commented:
I was under the impression that the `inline` keyword was implicitly applied to any functions defined inside the class definition. Do you actually get different results with and without those `inline`s?

noseglasses commented:
`DFU`'s methods can't be inlined right now. Implementations are AFAICS in the .cpp file.

obra commented:
In general, I like the direction of this. 

Why is the DFU implementation the only one protected by an ifdef? Can we make the DFU implementation more like the rest?

I feel like we should possibly be putting the platform name in the driver name. Something like AVRCaterina or AVR.Caterina or somesuch. I don't feel -strongly- about this.

algernon commented:
> Why is the DFU implementation the only one protected by an ifdef? Can we make the DFU implementation more like the rest?

Because DFU is the only one that needs protection, the rest can be compiled for boards that don't need them, and they will be left out by the linker. DFU - not so, because it uses a function that runs before `main()`, and as such, is always linked in. We don't want that linked in for non-DFU boards, because it would interfere with the normal boot. So sadly DFU has to be protected this way, and can't be like the rest.

(Did I mention DFU is bad? No? DFU is bad.)

>  feel like we should possibly be putting the platform name in the driver name. Something like AVRCaterina or AVR.Caterina or somesuch.

How about namespacing them instead? `kaleidoscope::driver::bootloader::avr::Caterina` & so on?

obra commented:
On Mon, Mar 18, 2019 at 5:07 PM Gergely Nagy <notifications@github.com>
wrote:

> Why is the DFU implementation the only one protected by an ifdef? Can we
> make the DFU implementation more like the rest?
>
> Because DFU is the only one that needs protection, the rest can be
> compiled for boards that don't need them, and they will be left out by the
> linker. DFU - not so, because it uses a function that runs before main(),
> and as such, is always linked in. We don't want that linked in for non-DFU
> boards, because it would interfere with the normal boot. So sadly DFU has
> to be protected this way, and can't be like the rest.
>
> (Did I mention DFU is bad? No? DFU is bad.)
>
>
I'm genuinely surprised that needs to happen there. Given how the reboot to
bootloader works, it ought to be possible to set all that just before the
wdt hackery. But if it genuinely needs to happen there, please document it
in comments.



> feel like we should possibly be putting the platform name in the driver
> name. Something like AVRCaterina or AVR.Caterina or somesuch.
>
> How about namespacing them instead?
> kaleidoscope::driver::bootloader::avr::Caterina & so on?
>

great.

> —
>



> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope/pull/611#issuecomment-474148001>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACxaLazQj4ZOe4BdGLhG88CQHoHn8cEks5vYCokgaJpZM4b33FL>
> .
>
ᐧ

algernon commented:
> I'm genuinely surprised that needs to happen there. Given how the reboot to bootloader works, it ought to be possible to set all that just before the wdt hackery. But if it genuinely needs to happen there, please document it in comments.

As far as I see, the issue is that unlike Caterina and HalfKay, the DFU bootloader doesn't support conditionally staying in bootloader mode. I mean, for Caterina, we can push a magic value to a magic location before forcing the watchdog to reboot us, the bootloader checks that, and doesn't start the firmware if the magic value is set. With HalfKay, we clear a bunch of stuff and directly jump to the bootloader, and if all that stuff is clear, it will stay there.

With DFU, we can reboot the keyboard, but the bootloader will start us up again anyway. We need a way to jump to the bootloader after a watchdog reset, and the bootloader itself does not offer a way to do that. So we hook into an section that runs before `main()`, and do the optional jump ourselves.

Mind you, reading the code now, I do wonder if we can pull a HalfKay and just jump directly to the bootloader after clearing the same stuff? I'll have a go at that. If that works, we can un-ifdef the DFU case. Which we should probably rename to FLIP, because that's the name of the bootloader, while DFU is the protocol to upload with. (Like AVR109 & Caterina)

In either case, I'll document all three resets in-code, with pointers to datasheets.

gedankenexperimenter commented on src/kaleidoscope/driver/Bootloader.h:
```diff
@@ -0,0 +1,99 @@
+ public:
+  Caterina() {}
+
+  static inline void resetDevice() {
```
I commented earlier about the use of `inline` here (and elsewhere) inside class definitions, but I was perhaps unclear. Perhaps I have misunderstood things that I've read, but my understanding was that all functions defined inside the class definition are implicitly `inline`, so the keyword has no effect here. Is there a specific purpose for using it here?

algernon replied:
Nope, no specific purpose. It's a leftover from an earlier attempt, and I'll be removing them. Thanks!

algernon commented:
- [x] Removed the `inline` keyword
- [x] Moved everything into the `::avr` namespace
- [x] Renamed `DFU` to `FLIP`, because FLIP is the bootloader, DFU is the protocol
- [ ] In-code documentation
- [ ] Developer documentation a'la the WS2812 driver

algernon commented:
Tested on the Model01 (Caterina), ErgoDox EZ (HalfKay) and Splitography (FLIP), and resetDevice() works on all of them.

obra commented:
Hm. Would it make more sense for each bootloader driver to be in its own file? shoving multiple implementations together feels a little not quite right to me.

algernon commented:
I can do that, too. The reason I kept them in a single file is because they're pretty small, and there are keyboards who can - in theory at least - use either. For example, there are Atreus builds with Teensy, and other builds with 32u4 MCUs that shipped with FLIP. We currently don't make it easy to choose between those, but once we do, we should be able to pull in the appropriate bootloader driver. If they're all in one file, I don't need to include two.

Mind you, if you want them split, I'll split 'em, no problem.

noseglasses commented:
> shoving multiple implementations together feels a little not quite right to me.

Allow me to express my opinion about this question. I am definitely with @obra here and really would opt for having only one public class per file and have filenames that coincide with class names, if possible. That makes browsing code so much easier. i find myself using grep on the Kaleidoscope code already far too often.

algernon commented:
A'ight, I'll make 'em split. Will push an update later tonight.

algernon commented:
Didn't get around to this last night, updated the PR now: the `Bootloader.h` header has been split, documentation updated too.

Additionally, the `FLIP.h` header will throw an error now when directly included without the `KALEIDOSCOPE_BOOTLOADER_FLIP` flag set. Previously, it just compiled an empty `resetDevice()` silently. I think the new behaviour is more correct, and less prone to errors.

obra commented on src/kaleidoscope/hardware/technomancy/Atreus.h:
```diff
@@ -65,9 +66,12 @@ class Atreus: public kaleidoscope::hardware::ATMegaKeyboard {
-  void resetDevice();
+  void resetDevice() {
+    bootloader_.resetDevice();
+  }
```
Is there a reason that things that extend ATMegaKeyboard can't use a shared implementation of resetDevice?

algernon replied:
Yes: Not all of them use the same bootloader. We could extend the `ATMegaKeyboard` macros to also take a bootloader arg, but I don't think that'd be worth it.

What we could do, on the other hand, is this:

```c++
#define KALEIDOSCOPE_USE_BOOTLOADER(x) \
 void resetDevice() { bootloader_.resetDevice(); } \ private: x bootloader_; public:
```

obra commented:
could we not to something to give an accessor that returns the bootloader
driver object?
ᐧ

On Wed, Mar 27, 2019 at 5:45 PM Gergely Nagy <notifications@github.com>
wrote:

> *@algernon* commented on this pull request.
> ------------------------------
>
> In src/kaleidoscope/hardware/technomancy/Atreus.h
> <https://github.com/keyboardio/Kaleidoscope/pull/611#discussion_r269823231>
> :
>
> > @@ -65,9 +66,12 @@ class Atreus: public kaleidoscope::hardware::ATMegaKeyboard {
>
>    static constexpr int8_t led_count = 0;
>
> -  void resetDevice();
> +  void resetDevice() {
> +    bootloader_.resetDevice();
> +  }
>
> Yes: Not all of them use the same bootloader. We could extend the
> ATMegaKeyboard macros to also take a bootloader arg, but I don't think
> that'd be worth it.
>
> What we could do, on the other hand, is this:
>
> #define KALEIDOSCOPE_USE_BOOTLOADER(x) \
>  void resetDevice() { bootloader_.resetDevice(); } \ private: x bootloader_; public:
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope/pull/611#discussion_r269823231>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACxaNpw5pTIpBWjsRfOW69zjQ-KX486ks5vbBCZgaJpZM4b33FL>
> .
>

algernon commented:
> could we not to something to give an accessor that returns the bootloader driver object?

So `ATMegaKeyboard` would have `void resetDevice() { bootloader_().resetDevice(); }`? How would we declare `bootloader_()`? We don't know the type yet, unless we add it to the macros, and AFAIK, we can't use `auto` there. We might be able to use a base class for the bootloaders, but then we have a lot more code than we have now. And we'd have to declare the bootloader _somewhere_, and it'd be specific to that instance of the `ATMegaKeyboard`-derivative, so it'd need to be in the header and .cpp file as well. We're no better than now, just papered over ~5 lines per derived board, with ~50 of "common" magic, and cut it down to ~2 lines / board. That doesn't seem like a good trade to me.

We could provide a default `resetDevice` that used, say, `Caterina` (since `resetDevice` is static on the bootloader drivers, we can do that without having a member variable, though then it's a bit uglier), but I don't see that as worthwhile either. Most ATMegaKeyboard-derivatives will ship with FLIP or HalfKay, rendering the default useless. We can't make FLIP the default because reasons explained above, and I'd rather not make HalfKay the default, because that one's not open source.

We COULD make `ATMegaKeyboard` templated, and then it could provide a `resetDevice`:

```c++
class SomeBoard: public kaleidoscope::hardware::ATMegaKeyboard<kaleidoscope::driver::bootloader::avr::FLIP> {
 // ...
}
```

This isn't _that_ bad, but I don't think we can make it backwards-compatible. Thankfully, in this case, that's no big deal, because `ATMegaKeyboard` is not user-facing, and there are no users of it that we don't control (at least, none that I know of).

obra commented:
I'm ok with breaking backcompat here.

Templated is appealing. Especially if we have a way to support options for
other drivers. I'm looking for us to establish good patterns we can reuse
for other things here.
ᐧ

On Wed, Mar 27, 2019 at 11:26 PM Gergely Nagy <notifications@github.com>
wrote:

> could we not to something to give an accessor that returns the bootloader
> driver object?
>
> So ATMegaKeyboard would have void resetDevice() {
> bootloader_().resetDevice(); }? How would we declare bootloader_()? We
> don't know the type yet, unless we add it to the macros, and AFAIK, we
> can't use auto there. We might be able to use a base class for the
> bootloaders, but then we have a lot more code than we have now. And we'd
> have to declare the bootloader *somewhere*, and it'd be specific to that
> instance of the ATMegaKeyboard-derivative, so it'd need to be in the
> header and .cpp file as well. We're no better than now, just papered over
> ~5 lines per derived board, with ~50 of "common" magic, and cut it down to
> ~2 lines / board. That doesn't seem like a good trade to me.
>
> We could provide a default resetDevice that used, say, Caterina (since
> resetDevice is static on the bootloader drivers, we can do that without
> having a member variable, though then it's a bit uglier), but I don't see
> that as worthwhile either. Most ATMegaKeyboard-derivatives will ship with
> FLIP or HalfKay, rendering the default useless. We can't make FLIP the
> default because reasons explained above, and I'd rather not make HalfKay
> the default, because that one's not open source.
>
> We COULD make ATMegaKeyboard templated, and then it could provide a
> resetDevice:
>
> class SomeBoard: public kaleidoscope::hardware::ATMegaKeyboard<kaleidoscope::driver::bootloader::avr::FLIP> {
>  // ...
> }
>
> This isn't *that* bad, but I don't think we can make it
> backwards-compatible. Thankfully, in this case, that's no big deal, because
> ATMegaKeyboard is not user-facing, and there are no users of it that we
> don't control (at least, none that I know of).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope/pull/611#issuecomment-477465819>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACxaALSLga3ZdMmTESsHiJEi_spMhLkks5vbGCqgaJpZM4b33FL>
> .
>

algernon commented:
What kind of options are you thinking of?

We can certainly do `public h:ATMegaKeyboard<d:b:a:WeirdThing<whatever>>`, and nest templates. We can _probably_ do `ATMegaKeyboard<WeirdThing(whatever)>` too, but I need to test that.

obra commented:
I was thinking of things like LED drivers and things we haven't thought of
yet ;)
ᐧ

On Wed, Mar 27, 2019 at 11:38 PM Gergely Nagy <notifications@github.com>
wrote:

> What kind of options are you thinking of?
>
> We can certainly do public h:ATMegaKeyboard<d:b:a:WeirdThing<whatever>>,
> and nest templates. We can *probably* do
> ATMegaKeyboard<WeirdThing(whatever)> too, but I need to test that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope/pull/611#issuecomment-477468363>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACxaPk5_1O6r5_zXThH9Z4rIx4pSOpyks5vbGNWgaJpZM4b33FL>
> .
>

algernon commented:
Aha, I see. We can always use more template parameters, and IIRC we can default to no-leds too. Will have to test, or summon @noseglasses and siphon his knowledge O:)

noseglasses commented:
i'm not sure it it is useful here, but I'd like to mention that there's an approach that is called [curiously recurring template pattern](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern). It can help in cases where a base class is meant to provide services that operate on members of derived classes. It's basically a polymorphism as all derived classes will inherit a common set of methods from their base class, but without virtual functions. That's why it's also called static polymorphism.

The basic idea is that a class is derived from a template class that is instanciated with the derived class as template parameter. Sounds weird, no?

it goes like this (i provide an example that uses the classes from this discussion, although i am not sure if any of it makes sense here).
```cpp
template<typename Derived__>
class ATMegaKeyboard {
   public:
      //...
      void resetDevice() { derived().bootloader_.resetDevice(); }
   protected:
      
      //...
      Derived__ &derived() { return static_cast<Derived__&>(*this); }
};

class BootloaderType {
    public:
       //...
       void resetDevice() {}
};

class SomeATMegaDerivedKeyboard : public ATMegaKeyboard<SomeATMegaDerivedKeyboard>
{
   friend class ATMegaKeyboard<SomeATMegaDerivedKeyboard>;

   //...

   private:
      BootloaderType bootloader_;
};

int main() {
    SomeATMegaDerivedKeyboard kbd;
    kbd.resetDevice();
}
```

The nice thing about this is that one does not need to have a large number of template arguments for the base class, like `template<typename Bootloader__, typename AnotherModule__, typename YetAnotherModule__> class ATMegaKeyboard`. Only one template parameter is necessary which is the derived class.

With this approach, if new modules similar to the bootloader would be added to derived classes, only the base class (here `ATMegaKeyboard`) would need to be adapted and gain new service methods like `resetDevice()`. The template signature remains unchanged.

There's even a way to let the base class check at compile time if a derived class has a data member or method of specific type and name. Yes, that's possible, the approach is called [SFINAE](https://jguegant.github.io/blogs/tech/sfinae-introduction.html). Brace before you read that link. it takes templates to the next level :smirk:. 

Using SFINAE, the template base class could do a compile time distinction of derived class and treat them differently. 

By this means, some keyboards could add new modules. The base class would check for these modules being present and then implement controlling code. This would make future changes backward compatible as no other derived classes are affected if one of the keyboards implements a new module (say a microwave) that needs to be reset.

algernon commented:
This is exactly the kind of information I wanted to summon you, thank you! :)

*algernon holes himself to read it all*

noseglasses commented:
_noseglasses retreats into his bottle_

noseglasses commented:
@algernon, if you happen to play around with the SFINAE stuff like 
```cpp

#include <type_traits>

// A small helper class
//
template<bool T>
class Bool2Type {};

template<typename Derived__>
class ATMegaKeyboard {
   public:

      template <class T> struct hasMicrowave
      {
         // We test if the type has serialize using decltype and declval.
         template <typename C> static constexpr decltype(std::declval<C>().microwave_, bool()) test(int /* unused */)
         {
            // We can return values, thanks to constexpr instead of playing with sizeof.
            return true;
         }

         template <typename C> static constexpr bool test(...)
         {
            return false;
         }

         // int is used to give the precedence!
         static constexpr bool value = test<T>(int());
      };

      static constexpr bool derived_has_microwave = hasMicrowave<Derived__>::value;

      void heatStuff() { 
          this->heatStuff(Bool2Type<derived_has_microwave>());
      }

      // These methods are there to enable compile time selection
      // of the way to deal with microwave and non-microwave keyboards
      //
      void heatStuff(Bool2Type<true>) {
          derived().microwave_.heatStuff();
      }

      void heatStuff(Bool2Type<false>) {
          // A noop for those derived keyboards that come without microwave
      }

   protected:

      Derived__ &derived() { return static_cast<Derived__&>(*this); }
};

class Microwave {
    public:

       void heatStuff() {}
};

class SomeMicrowaveKeyboard : public ATMegaKeyboard<SomeMicrowaveKeyboard>
{
   friend class ATMegaKeyboard<SomeMicrowaveKeyboard>;

   private:
      Microwave microwave_;
};

class SomeNonMicrowaveKeyboard : public ATMegaKeyboard<SomeNonMicrowaveKeyboard>
{
   friend class ATMegaKeyboard<SomeNonMicrowaveKeyboard>;
};

class NonBootloaderKeyboard {};

int main() {
    SomeMicrowaveKeyboard mwk;
    mwk.heatStuff();

    SomeNonMicrowaveKeyboard nmwk;
    nmwk.heatStuff(); // noop
}
```

you will notice, that the `std::declval` function is missing. Normally, it lives in header `type_traits` which is not available as avr-gcc comes without a standard library. 

Fortunately, this header is kind-of independent from the rest of the standard lib. It's easy to steal a version from another complete gcc installation and fix it a little. I have done that: https://gist.github.com/noseglasses/d17ca459fb1931c8794d28aad4cd44b5#file-avr-g-_type_traits-h
I have tested it on compiler explorer with avr-gcc 4.6.4 and it appears to work correctly.

Type traits make things a lot easier when dealing with templates. i could well imagine to add our own version of the `type_traits` header to Kaleidoscope until avr-gcc ships with a standard lib one day.

What's your opinion?

algernon commented:
Oooh, I wanted to steal `type_traits` before, but never got around to it. It does make a lot of things nicer, I'm perfectly ok with shipping our version until avr-gcc starts shipping one.

algernon commented:
Having digested most of the information here, I'll go with the simpler way for now, without SFINAE. The method looks incredibly useful, and I see a lot of potential for it within Kaleidoscope, but I'd like to introduce that separately, primarily because I feel I need to understand it a bit better first. By the looks of it, derived classes won't need to change for that, so later introduction should be safe.

algernon commented:
Interesting... I migrated over ATMegaKeyboard to use the derived class as a template argument, it worked out quite fine, but we ended up with about 38-40 bytes more PROGMEM used. I'm not sure why yet... `elf_diff` tells me a bunch of symbols disappeared, others got renamed (as expected, we went from `ATMegaKeyboard` to `ATMegaKeyboard<foo>` after all), but the sizes of those don't suggest a PROGMEM increase.

I _think_ this is because everything was moved to the header, thus, the compiler can optimize more aggressively: some functions are marked for -O3, after all. If I remove the -O3 attributes, we see quite a size drop.

algernon commented:
Updated, `ATMegaKeyboard` is now templated, and uses the curiously recurring template pattern. I'm kinda liking this setup, it's much easier to extend. Still have to figure out why our PROGMEM size increases by ~40 bytes, and still need to test if things still work for `ATMegaKeyboard`-derived boards, so leaving it WIP for now.

noseglasses commented:
Have you checked the _similar symbols_ section that i recently added to elf_diff? It attempts to determine pairs of symbols that appear to be related and presents them in an additional table. This is especially useful when working on changes that involve renaming classes, functions, namespaces, etc.

You can click on the integer id of the similarity pair in the first column of the table and it will take you directly to a diff of the symbols. [in general when clicking on symbols you can jump back and forth between tables and diffs].

But it helped a lot when I was working on the linear key addressing topic where i also was confronted with a strange raise of PROGMEM consumption.

algernon commented:
I don't have the table in front of me, but IIRC it had no similar symbols. I'm working on another topic that makes use of the same pattern, and I'm seeing substantial PROGMEM use there, too. I'll post more info as soon as I have any.

noseglasses commented:
> I don't have the table in front of me, but IIRC it had no similar symbols. 

That's something I added not more than a week ago. I added an additional table and an additional set of diffs that represent something like "symbols that have likely undergone some sort of transformation but are worth being diffed".

algernon commented:
`elf_diff` (using elf_diff master) between Kaleidoscope `master` and the `driver/bootloader` branch, for the `Devices/KBD4x` example attached.

[master-bootloader-diff.html.txt](https://github.com/keyboardio/Kaleidoscope/files/3047110/master-bootloader-diff.html.txt)

noseglasses commented:
Thanks for sharing. I will see why the changes are not listed as _similar symbols_.

noseglasses commented:
While i was working on a fix for elf_diff, I noticed that your size increase is likely due to the compiler inlining the `bitRead` and `bitWrite` calls in `handleKeyswitchEvent`.

Here's a report that I created after I fixed elf_diff: 
[master-bootloader-diff-elf_diff-fixed.html.txt](https://github.com/keyboardio/Kaleidoscope/files/3047361/master-bootloader-diff-elf_diff-fixed.html.txt)

BTW, you elf_diff output lacked the actual binary diff information likely because it used the x86 versions of nm and objdump. To make it use the avr-versions, you must pass the parameters
```
elf_diff --bin-dir <path_to_arduino>/hardware/tools/avr/bin --bin-prefix "avr-" <binaries...>
```
I will let you know once i pushed the latest changes to elf_diff.

noseglasses commented:
One more thing about the binary diffs. Objdump outputs six lines of source context at the start of every assembler block. That's a bit hard to read because I did not find an easy way yet to provide different highlighting for source code and assembler output in HTML, yet.

algernon commented:
Aha, that makes sense. Thanks for looking into this!

noseglasses commented:
I just pushed a fix for elf_diff. If you find any other peculiarities, please submit an issue against the elf_diff repo.

gedankenexperimenter commented:
Frustratingly, none of the web browsers on my phone is willing and able to render that file as HTML. Most of them just display the source as text, even if the `.txt` extension is removed, and Firefox just crashes outright…

noseglasses commented:
It's not meant to be used like that. It's lacking a js file that lies in the elf_diff repo. Just copy it there and it will work. 

Mind you, elf_diff is still work in progress.

obra commented on src/kaleidoscope/hardware/ATMegaKeyboard.h:
```diff
@@ -67,44 +67,182 @@ struct cRGB {
   uint8_t BOARD::debounce_matrix_[matrix_rows][matrix_columns]; \
```
Should 'BOARD' here change to KeyboardHardware?

algernon replied:
No, because we need the class name here, not the instance name.

obra commented:
@algernon - is this ready for merge?

algernon commented:
No, not yet.

obra commented:
@algernon - what needs to happen to get this to mergeable?

algernon commented:
I need to rebase it on top of master, test it, and either figure out why we see a PROGMEM increase on some boards (ATMegaKeyboard-based ones) or just ignore that and live with it.

algernon commented:
I have rebased _most_ of this on top of master, but the linear addressing stuff broke the ATMegaKeyboard-related changes in this PR so much that I've yet to unbreak it. I'm pushing my rebased branch nevertheless, so others can take a look too.

algernon commented:
So, the problem is, that `ATMegaKeyboard` would need to know the explicit `KeyAddr` of the derived keyboard, but that kinda creates a dependency loop. Not quite sure how to get around that. Unless we split up `ATMegaKeyboard`, so we only have to template a part of it, the one that doesn't need this info.

obra commented:
@algernon Can you explain a bit more about what's wrong? I don't quite get it from your note

algernon commented:
Right, sorry. So, the problem is that we turned `ATMegaKeyboard` into a templated class. That means, it had to be moved to the header, fully. And that's a problem, because `ATMegaKeyboard::actOnMatrixScan` calls `handleKeyswitchEvent`, which takes a `KeyAddr` argument. That, however, is hardware specific by design, and `ATMegaKeyboard` does not have access to it.

This means that the _current_ branch I have pushed, doesn't compile, because `KeyAddr` doesn't resolve. I can't make it resolve, because `ATMegaKeyboard` can't depend on the derived keyboard, and while I can access _members_ of the derived board via `derivedKeyboard_`, I can't access typedefs.

To solve this, either we need to rework the bootloader stuff, in a way that `ATMegaKeyboard` doesn't have to be templated, or split it up in a way that `actOnMatrixScan` is _not_ templated.

Does this make more sense?

noseglasses commented:
I've added a proposal for an slightly alternative way to modularize the keyboard hardware and that deals with the problems described above. Please find it here: https://github.com/keyboardio/Kaleidoscope/issues/665

algernon commented:
I've remade this on top of #695. Once that is done and merged, I'll rebase this again on top of master.

algernon commented:
For ease of development, I folded this into #695. We may end up removing this part from there, at which point I will update the PR. But for the latest working code, see #695 for now.

algernon commented:
This has been folded into #695 completely. Closing this one.
